src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp

```diff
@@ -67,21 +67,21 @@ void smt2_incremental_decision_proceduret::define_dependent_functions(
     {
       const irep_idt &identifier = symbol_expr->get_identifier();
       const symbolt *symbol = nullptr;
-      if(!ns.lookup(identifier, symbol) && !symbol->value.is_nil())
+      if(ns.lookup(identifier, symbol) || symbol->value.is_nil())
```

Comment:
Are we sure this change is valid? It doesn't look like that to me.

[De Morgan's law in this particular instance would be a bi-implication](https://en.wikipedia.org/wiki/De_Morgan%27s_laws):

```
¬(A ∨ B) ⇔ ¬A ∧ ¬B
```

Assuming in your case `A : ns.lookup()` and `B : symbol->value.is_nil()` then I believe this should have been translated to `(!(ns.lookup() || symbol->value.is_nil()))`.

Reply:
Yeah, I just verified this is with a quick run of the truth table:

`A : ns.lookup()`
`B : symbol->value.is_nil()`
`Original : !ns.lookup() && !symbol->value.is_nil()`
`Refactored : ns.lookup() || symbol->value.is_nil()`

| A  | B | Original | Refactored |
|---|---|--------|-------------|
| F  | F  |      T      |         F           |
| F  | T  |       F     |         T           |
| T  | F  |       F     |         T           |
| T  | T  |       F     |        T            |

Reply:
Yes. The outcome of the logic you present is the opposite way round. That final not was handled by flipping the contents of the code blocks rather than adding the `not` operation. Therefore the outcome is still the same, which is why the tests still pass.